### PR TITLE
[TT-XLA][PyTorch] GRPO Training Workload for Gemma 2 2B Instruct - GSM8K

### LINKED ISSUE #674: GRPO Training on GSM8K
### Is your feature request related to a problem? Please describe

No

### Describe the solution you'd like

We want to add training of Gemma 2 2B Instruct using GRPO like in the blog https://medium.com/@lucamassaron/training-for-reasoning-with-grpo-part-ii-a-step-by-step-explanation-f80c219e2059

### PR DESCRIPTION
**Pull request by mmilosevicTT**
### Issue
Closes https://github.com/tenstorrent/tt-blacksmith/issues/674

### Training Workload Setting

- **Model:** Gemma 2 2B Instruct
- **Strategy:** GRPO + LoRA
- **Framework:** PyTorch
- **Task:** Question Answering
- **Dataset:** GSM8K
- **Hardware:** P150

### What's Changed
This PR showcases training of Gemma 2 2B Instruct on GSM8K using GRPO method. For reward we were using following:
- Format Scoring (`<reasoning>...</reasoning><answer>...</answer>`) 0 or 1 point
- Correctness Scoring (number between <answer> tags, or last number in completion) 0 or 2 points

This training is bound to the fix in https://github.com/tenstorrent/tt-xla/issues/5793. The decode bug block us from running proper training, as compilation is really slow.

### Experiment Results
Here, we are providing results of the GRPO run on singlechip:

<table>
  <tr>
    <th>GRPO Training Loss</th>
    <th>GRPO Training Reward</th>
  </tr>
  <tr>
    <td align="center">
loss_final
    </td>
    <td align="center">
reward_final
    </td>
  </tr>
  <tr>
    <th>GRPO Training Format Reward Fraction</th>
    <th>GRPO Training Correctness Reward Fraction</th>
  </tr>
  <tr>
    <td align="center">
format_final
    </td>
    <td align="center">
correct_final
    </td>
  </tr>
</table>

From the results, we can see that the model quickly learns the format part of the reward, but that it struggles learning correctness. This can be due to expressiveness of the model, as it only has 2B parameters and is trained with LoRA adapters only.

### Checklist
- [x] Main training workflow
- [x] GPU testing
- [x] TT device testing

### CODE REVIEW

File: blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+# SPDX-License-Identifier: Apache-2.0
+"""
+GSM8K dataset for GRPO training.
+
+Unlike the supervised datasets, GRPO generates its own completions at train
+time, so this dataset is PROMPT-ONLY: each item is a chat-templated, R1-style
+prompt plus the gold integer answer used by the correctness reward. Prompts are
+left-padded at collate time so batched autoregressive generation shares a single
+generation frontier.
+"""
```
Comment (pglusacTT) [inline blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py:6]: This comment can be removed, wdyt?

File: blacksmith/experiments/torch/gemma2/grpo/train.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+                # Keep one decoded completion (a plain string) for logging before the
+                # underlying tensors are freed below.
+                sample_completion = completions_text[0] if completions_text else ""
+
+                # Explicitly drop per-iteration tensors so they don't pin device/host
+                # memory across the gradient-accumulation window. Placed before the
+                # early-continue below so it runs on every iteration. The metrics were
+                # already accumulated (as detached scalars) just above.
+                #TODO del prompt_ids, prompt_mask, golds
```
Comment (pglusacTT) [inline blacksmith/experiments/torch/gemma2/grpo/train.py]: TODOs should have a username attached to them.

File: blacksmith/tools/grpo_utils.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+from typing import Dict, List, Tuple
+
+import torch
+import torch.nn.functional as F
+
+from blacksmith.datasets.torch.gsm8k.gsm8k_utils import extract_predicted_answer
+
+# A completion is well-formatted when a reasoning block is immediately followed
+# by an answer block.
+_FORMAT_PATTERN = re.compile(r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>", re.DOTALL)
```
Comment (pglusacTT) [inline blacksmith/tools/grpo_utils.py]: I know that this pattern is common but isn't it a bit too model/dataset specific to be located inside of `tools/` ?

File: blacksmith/experiments/torch/gemma2/grpo/train.py (hunk 2 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma2/grpo/train.py]: Good catch. I actually want to test out this before merging. Left it as a reminder and forgot. Thanks for the ping.

File: blacksmith/tools/grpo_utils.py (hunk 3 above)
Comment (mmilosevicTT) [reply blacksmith/tools/grpo_utils.py]: true, it might be worth moving it with dataset utils. Will move.

File: blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py (hunk 1 above)
Comment (mmilosevicTT) [reply blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py:6]: Agreed, will remove.

File: blacksmith/datasets/torch/gsm8k/gsm8k_utils.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+def format_reward(completion: str) -> float:
+    """1.0 if the completion matches the reasoning/answer tag format, else 0.0."""
+    return 1.0 if _FORMAT_PATTERN.search(completion) else 0.0
+
+
+def correctness_reward(completion: str, gold: str) -> float:
+    """1.0 if the extracted answer equals the gold answer, else 0.0."""
+    pred = extract_predicted_answer(completion)
+    return 1.0 if pred != "" and pred == gold else 0.0
```
Comment (umalesTT) [inline blacksmith/datasets/torch/gsm8k/gsm8k_utils.py]: Can't we just do `1.0 * (pred == gold)`?

File: blacksmith/experiments/torch/gemma2/grpo/configs.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    ignored_index: int = Field(default=-100)
+
+    # GRPO / generation hyperparameters
+    num_generations: int = Field(
+        default=4, gt=1, description="G: completions sampled per prompt (>=2 for advantage std)"
+    )
+    num_iterations: int = Field(
+        default=1,
+        ge=1,
+        description="μ: policy updates on the same rollouts before refreshing pi_old",
```
Comment (umalesTT) [inline blacksmith/experiments/torch/gemma2/grpo/configs.py]: Nit (and up for discussion): I'd really use terms like old/new policy model, reference model, instead of symbolic pi_old

File: blacksmith/experiments/torch/gemma2/grpo/train.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    optimizer = torch.optim.AdamW(
+        [p for p in policy_model.parameters() if p.requires_grad],
+        lr=config.learning_rate,
+        weight_decay=config.weight_decay,
+        capturable=config.use_tt,
+    )
+    optimizer.zero_grad()
+
+    # Per-step device sync used to bound the generation graph.
+    generation_sync = (lambda: torch_xla.sync(wait=True)) if config.use_tt else None
```
Comment (umalesTT) [inline blacksmith/experiments/torch/gemma2/grpo/train.py]: This is used only in `generate_completions`, right? Why not just go with "if" as it is elsewhere?

**Review by umalesTT — APPROVED** [review]
LGTM 🔥 
Were you rebased on https://github.com/tenstorrent/tt-xla/pull/5771 while running this, because I remember offline we spoke about prefill<->decode issues?

File: blacksmith/datasets/torch/gsm8k/gsm8k_utils.py (hunk 4 above)
Comment (mmilosevicTT) [reply blacksmith/datasets/torch/gsm8k/gsm8k_utils.py]: Agreed. Fixed.

File: blacksmith/experiments/torch/gemma2/grpo/configs.py (hunk 5 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma2/grpo/configs.py]: Agreed. It is much clearer to keep it in text then to use pi.

File: blacksmith/experiments/torch/gemma2/grpo/train.py (hunk 6 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma2/grpo/train.py]: My idea here was not to expose torch_xla import in the troch_utils, but it is not a big deal. Changed.

**Comment by mmilosevicTT** [comment]
> LGTM 🔥 Were you rebased on [tenstorrent/tt-xla#5771](https://github.com/tenstorrent/tt-xla/pull/5771) while running this, because I remember offline we spoke about prefill<->decode issues?

I cherry-picked the commit, but it was not sufficient. Will open an issue and link it to this PR.

File: blacksmith/tools/logging_manager.py [hunk 7]
```diff
@@ -107,8 +107,13 @@ def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None, commi
                 self.std_logger.warning(f"Failed to log to W&B: {e}")
 
         if self.test_log_filename_prefix is not None:
+            row = {"_step": step}
             if "train/loss" in metrics:
-                self.train_log.append({"_step": step, "train/loss": metrics["train/loss"]})
+                row["train/loss"] = metrics["train/loss"]
+            if "train/reward_mean" in metrics:
+                row["train/reward_mean"] = metrics["train/reward_mean"]
+            if len(row) > 1:
```
Comment (agobeljicTT) [inline blacksmith/tools/logging_manager.py]: this gonna be > 1 always, you have row = {step: 1}

File: blacksmith/tools/torch_helpers.py [hunk 8]
```diff
@@ -249,6 +250,157 @@ def run_decode_example(
...
+                cache_position=cache_position,
+                use_cache=True,
+            )
+            next_logits = output.logits[:, -1, :]
+            # A row's token this step is valid iff it had not already finished.
+            valid_this = ~finished
+            next_tokens = _sample_next_tokens(next_logits, temperature, top_k, sample_rng_on_cpu=sample_rng_on_cpu)
+            # Force already-finished rows to pad so generation stays batched.
+            next_tokens = torch.where(finished, pad_id.expand_as(next_tokens), next_tokens)
+            finished = finished | (next_tokens == eos_id)
```
Comment (agobeljicTT) [inline blacksmith/tools/torch_helpers.py]: some models use <end_of_turn>, are we safe with this check alone?

File: blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+from datasets import load_dataset
+
+
+class GSM8KDataset(BaseDataset):
+    """Prompt-only GSM8K dataset yielding ``{"prompt", "gold"}`` items."""
+
+    def __init__(self, config, split: str = "train", collate_fn=None):
+        # Left padding is required for batched generation: it aligns the last real
+        # prompt token to a common index across the batch.
+        self.tokenizer = AutoTokenizer.from_pretrained(config.model_name, padding_side="left", use_fast=True)
```
Comment (agobeljicTT) [inline blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py:29]: truncation side is right, but padding side is left
do we need to align this?

File: blacksmith/experiments/torch/gemma2/grpo/train.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+                # the new-policy forward's activations that must stay resident for
+                # backward.
+                old_logps = compute_old_logps(old_policy_model, seq_ids, seq_attention_mask)
+                ref_logps = compute_ref_logps(
+                    policy_model,
+                    seq_ids,
+                    seq_attention_mask,
+                    use_shared_reference=use_shared_reference,
+                    reference_model=reference_model,
+                )
```
Comment (agobeljicTT) [inline blacksmith/experiments/torch/gemma2/grpo/train.py:268]: what happens when μ=1?
old_logps should be equal to logps.detach(), right?

File: tests/golden_files/tt-gemma2-gsm8k_grpo-p150_train.csv [hunk 11]
```diff
@@ -0,0 +1,5 @@
+_step,train/loss,train/reward_mean
+1,0,3
+2,0,1.5
```
Comment (agobeljicTT) [inline tests/golden_files/tt-gemma2-gsm8k_grpo-p150_train.csv]: loss begins with 0?

File: tests/test_all.py [hunk 12]
```diff
@@ -95,19 +102,38 @@ def run_cmd(cmd: list[str], test_id: str, setup_dict: dict, debug: bool):
...
-        GOLDEN_LOGS_DIR / val_log_file,
+def check_metrics(train_log_file: Path, val_log_file: Path, setup_dict: dict):
+    train_log = TEST_LOGS_DIR / train_log_file
+    train_golden = GOLDEN_LOGS_DIR / train_log_file
+    assert_metric_with_tolerance(
+        train_log,
+        train_golden,
+        "train/loss",
         tolerance=setup_dict["tolerance"],
         atol=setup_dict["atol"],
```
Comment (agobeljicTT) [inline tests/test_all.py:113]: perhaps a good time to rename this to rtol?

File: tests/test_all.py [hunk 13]
```diff
@@ -160,8 +186,11 @@ def test_training_script(
     if request.config.getoption("--generate-golden-files"):
         # Reference run, move the log files to golden_files.
         (TEST_LOGS_DIR / train_log_file).rename(GOLDEN_LOGS_DIR / train_log_file)
-        (TEST_LOGS_DIR / val_log_file).rename(GOLDEN_LOGS_DIR / val_log_file)
+        val_src = TEST_LOGS_DIR / val_log_file
+        golden_val = GOLDEN_LOGS_DIR / val_log_file
+        if val_src.exists():
+            val_src.rename(golden_val)
```
Comment (agobeljicTT) [inline tests/test_all.py:192]: I don't like this, we should always check for val in the CI.
for local runs we can just ignore these failures

File: tests/training_test_cases.py [hunk 14]
```diff
@@ -358,6 +358,27 @@
...
+                reason="tt-xla: non-AOT StaticCache decode packs KV mutations as logits.",
+                strict=False,
+            ),
+            pytest.mark.uplift,
+            pytest.mark.p150,
+            pytest.mark.torch,
+            pytest.mark.single_chip,
+        ],
+        id="tt-gemma2-gsm8k_grpo-p150",
+    ),
```
Comment (agobeljicTT) [inline tests/training_test_cases.py:381]: make sure this runs fine by running the uplift job on this PR

File: blacksmith/tools/grpo_utils.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    The (B, T, V) logits are dropped immediately after reduction to keep only the
+    small per-token tensor (avoids holding a vocab-sized activation).
+    """
+    logits = model(input_ids=seq_ids, attention_mask=seq_attention_mask).logits
+    return get_per_token_logps(logits, seq_ids)
+
+
+def sync_old_policy(policy_model: torch.nn.Module, old_policy_model: torch.nn.Module) -> None:
+    """Copy new-policy weights into the frozen old-policy model."""
+    old_policy_model.load_state_dict(policy_model.state_dict())
```
Comment (agobeljicTT) [inline blacksmith/tools/grpo_utils.py]: I think this copies the entire state dict. We can save some time by sharing refs or just copying trainable params.

File: blacksmith/experiments/torch/gemma2/grpo/README.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+### Configuration Parameters
+
+| Parameter | Description | Default Value |
+| --- | --- | --- |
+| `training_model_type` | Fine-tuning approach for the policy model (`lora`, `adapters`, or full). | "lora" |
+| `dataset_id` | The dataset used for training. | "gsm8k" |
+| `model_name` | Name or path of the pre-trained Gemma 2 2B model. | "google/gemma-2-2b-it" |
+| `dtype` | Data type used during training. | "torch.bfloat16" |
+| `ignored_index` | Label id used to mask prompt tokens. | -100 |
+| `num_generations` | G: completions sampled per prompt (>=2). | 4 |
```
Comment (agobeljicTT) [inline blacksmith/experiments/torch/gemma2/grpo/README.md]: this is by default 1? sync readme with defaults

File: blacksmith/experiments/torch/gemma2/grpo/README.md [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+| `checkpoint_metric` | Metric monitored for best checkpoint. | "train/reward_mean" |
+| `checkpoint_metric_mode` | Mode for checkpoint metric (`min` or `max`). | "max" |
+| `keep_last_n` | Number of recent checkpoints to keep. | 3 |
+| `keep_best_n` | Number of best checkpoints to keep. | 1 |
+| `save_strategy` | Strategy for saving checkpoints (`epoch`, `step`, or `none`). | "step" |
+| `save_steps` | Frequency (in steps) for saving checkpoints. | 50 |
+| `project_dir` | Directory for experiment outputs. | "blacksmith/experiments/torch/gemma2/grpo" |
+| `seed` | Random seed for reproducibility. | 23 |
+| `deterministic` | Whether to enforce deterministic behavior. | False |
+| `mesh_shape` | Mesh shape for distributed training. | None |
```
Comment (agobeljicTT) [inline blacksmith/experiments/torch/gemma2/grpo/README.md]: I don't see this is used, perhaps missed something...

File: tests/training_test_cases.py (hunk 14 above)
Comment (mmilosevicTT) [reply tests/training_test_cases.py:381]: https://github.com/tenstorrent/tt-blacksmith/actions/runs/30802145932 did this, but need to wait 😄

File: tests/test_all.py (hunk 13 above)
Comment (mmilosevicTT) [reply tests/test_all.py:192]: validation is not done during training for GRPO, so we cannot check it.

File: tests/golden_files/tt-gemma2-gsm8k_grpo-p150_train.csv (hunk 11 above)
Comment (mmilosevicTT) [reply tests/golden_files/tt-gemma2-gsm8k_grpo-p150_train.csv]: Yes, this is expected. The loss is not done in the same way as for LLM fine-tuning with LoRA. Policy is same as reference so the reward ends up being advantage only and expectation of advantage is 0, since it is normalized.

File: blacksmith/experiments/torch/gemma2/grpo/train.py (hunk 10 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma2/grpo/train.py:268]: yes, and that is expected.

File: blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py (hunk 9 above)
Comment (mmilosevicTT) [reply blacksmith/datasets/torch/gsm8k/gsm8k_dataset.py:29]: Yeah, you are right. We do not want to cut off the example at all.

File: blacksmith/tools/logging_manager.py (hunk 7 above)
Comment (mmilosevicTT) [reply blacksmith/tools/logging_manager.py]: Not really, it will be 1 if we miss loss/reward metrics, this forces the experiment to log loss in the same way, but I agree we do not need a check.

File: tests/test_all.py (hunk 12 above)
Comment (mmilosevicTT) [reply tests/test_all.py:113]: Well, I wouldn't mix renamings with GRPO PR as the PR is complex in it's nature already.

File: tests/training_test_cases.py (hunk 14 above)
Comment (mmilosevicTT) [reply tests/training_test_cases.py:381]: https://github.com/tenstorrent/tt-blacksmith/actions/runs/30841175186/job/91778361541
